velox/functions/sparksql/String.h

```diff
@@ -283,6 +283,80 @@ struct EndsWithFunction {
   }
 };
 
+/// substring_index function
+/// substring_index(string, string, int) -> string
+/// substring_index(str, delim, count) - Returns the substring from str before
+/// count occurrences of the delimiter delim. If count is positive, everything
+/// to the left of the final delimiter (counting from the left) is returned. If
+/// count is negative, everything to the right of the final delimiter (counting
+/// from the right) is returned. The function substring_index performs a
+/// case-sensitive match when searching for delim.
+template <typename T>
+struct SubstringIndexFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& str,
+      const arg_type<Varchar>& delim,
+      const int32_t& count) {
+    doCall<false>(result, str, delim, count);
+  }
+
+  FOLLY_ALWAYS_INLINE void callAscii(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& str,
+      const arg_type<Varchar>& delim,
+      const int32_t& count) {
+    doCall<true>(result, str, delim, count);
+  }
+
+  template <bool isAscii>
+  FOLLY_ALWAYS_INLINE void doCall(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& str,
+      const arg_type<Varchar>& delim,
+      const int32_t& count) {
+    if (count == 0) {
+      result.setEmpty();
+      return;
+    }
+
+    auto index = -1;
```

Comment:
Looks no need to initialize index as -1.
`int64_t index;`

Reply:
Done.